process/slash/detector/multipleHeaderProposalsDetector.go

```diff
@@ -0,0 +1,239 @@
+package detector
+
+import (
+	"bytes"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/block/interceptedBlocks"
+	"github.com/ElrondNetwork/elrond-go/process/slash"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+type detectorCache interface {
+	add(round uint64, pubKey []byte, data process.InterceptedData)
+	proposedData(round uint64, pubKey []byte) dataList
+}
+
+const CacheSize = 10
+const MaxDeltaToCurrentRound = 3
+const MinSlashableNoOfHeaders = 2
+
+// multipleHeaderProposalsDetector - checks slashable events in case a validator proposes multiple(possibly) malicious headers.
+type multipleHeaderProposalsDetector struct {
+	cache            detectorCache
+	nodesCoordinator sharding.NodesCoordinator
+	roundHandler     process.RoundHandler
+}
+
+// NewMultipleHeaderProposalsDetector - creates a new multipleHeaderProposalsDetector for multiple headers
+// proposal detection or multiple headers proposal proof verification
+func NewMultipleHeaderProposalsDetector(
+	nodesCoordinator sharding.NodesCoordinator,
+	roundHandler process.RoundHandler,
+	maxRoundCacheSize uint64,
+) (slash.SlashingDetector, error) {
+	if check.IfNil(nodesCoordinator) {
+		return nil, process.ErrNilShardCoordinator
+	}
+	if check.IfNil(roundHandler) {
+		return nil, process.ErrNilRoundHandler
+	}
+
+	//TODO: Here, instead of CacheSize, use maxRoundCacheSize = from config file
+	cache := newRoundProposerDataCache(CacheSize)
+	return &multipleHeaderProposalsDetector{
+		cache:            cache,
+		nodesCoordinator: nodesCoordinator,
+		roundHandler:     roundHandler,
+	}, nil
+}
+
+// VerifyData - checks if an intercepted data(which should be a header) represents a slashable event.
+// If another header with the same round and proposer exists, but a different hash, then a proof of type
+// slash.MultipleProposal is provided, otherwise a nil proof, along with an error is provided indicating that
+// no slashing event has been detected or an error occurred verifying the data.
+func (mhp *multipleHeaderProposalsDetector) VerifyData(data process.InterceptedData) (slash.SlashingProofHandler, error) {
+	header, castOk := data.(*interceptedBlocks.InterceptedHeader)
```

Comment:
can you rename here header to interceptedHeader?
and have also a header := interceptedHeader.HeaderHandler()

Reply:
Yes. It's better this way !
I've applied your suggestion